File: Python/src/seguid/asserts.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from seguid.tables import COMPLEMENT_TABLE
import seguid.manip

def assert_in_alphabet(seq: str,
                       alphabet: set):

    assert isinstance(seq, str), "Argument 'seq' must be an string"
    assert isinstance(alphabet, set), "Argument 'alphabet' must be a set"
    assert len(alphabet) > 0, "Argument 'alphabet' must not be empty"
    # breakpoint()
    # Nothing to do?
    if len(seq) == 0:
        return

    first = list(alphabet)[0]
    if isinstance(first, int):
        unknown = set(
            c for c in seq if c not in (chr(k) for k in alphabet)
        )
    elif isinstance(first, str):
        unknown = set(
            c for c in seq if c not in (k for k in alphabet)
        )
    else:
        raise ValueError("Unknown type of the elements in 'alphabet'")

    if unknown:
        missing = ' '.join(unknown)
        raise ValueError(
            "Detected symbols " f"{missing} in not in the 'alphabet'"
        )


def assert_table(table: dict):
    assert isinstance(table, dict), "Argument 'table' must be a dict"
    keys  = table.keys()
    values = table.values()
    # Assert that the set of values are also in the set of keys
    unknown = set(
        chr(v) for v in values if v not in (k for k in keys)
    )

    if unknown:
        missing = ' '.join(unknown)
        raise ValueError(
            "Detected values (" f"{missing}) in 'table' that are not in the keys"
        )


def assert_anneal(watson: str,
                  crick: str,
                  overhang: int,
                  table: dict = COMPLEMENT_TABLE) -> bool:
    """docstring."""
    assert_table(table)
    assert_in_alphabet(watson, alphabet=set(table.keys()))
    assert_in_alphabet(crick, alphabet=set(table.keys()))

    assert isinstance(overhang, int)
    assert -len(watson) < overhang
    assert overhang < len(crick)

    up = watson[max(-overhang, 0): min(len(watson) - overhang, len(crick))]
    dn = seguid.manip.rc(crick, table=table)[max(overhang, 0): min(len(watson) + overhang, len(crick))]

    if up != dn:
        raise ValueError("Mismatched basepairs.")

File: r/R/asserts.R
escape_symbols <- function(s) {
  for (kk in seq_along(s)) {
    ch <- s[kk]
    if (grepl("^[[:space:]]$", ch)) {
      ch <- gsub('(^"|"$)', "", deparse(ch))
      s[kk] <- ch;
    }
  }
  s
}

assert_valid_alphabet <- function(alphabet) {
  stopifnot(length(alphabet) > 0, is.character(alphabet), !anyNA(alphabet), all(nchar(alphabet) == 1))
  
  ## Allow only for 0-9, A-Z, a-Z, '-', '\n'
  unknown <- setdiff(alphabet, c(0:9, LETTERS, letters, "-", "\n"))
  if (length(unknown) > 0) {
    unknown <- escape_symbols(unique(sort(unknown)))
    stop(sprintf("Non-supported symbols in alphabet: [n=%d] %s", length(unknown), paste(sQuote(unknown), collapse = ", ")))
  }  
}  

assert_in_alphabet <- function(seq, alphabet) {
  stopifnot(length(seq) == 1, is.character(seq), !is.na(seq))
  assert_valid_alphabet(alphabet)

  ## Nothing to do?
  if (nchar(seq) == 0) {
    return(seq)
  }

  seq <- strsplit(seq, split = "", fixed = TRUE)[[1]]
  unknown <- setdiff(seq, alphabet)
  if (length(unknown) > 0) {
    unknown <- escape_symbols(unique(sort(unknown)))
    stop(sprintf("Sequence symbols not in alphabet: [n=%d] %s (not in %s)", length(unknown), paste(sQuote(unknown), collapse = ", "), paste(sQuote(escape_symbols(alphabet)), collapse = ", ")))
  }
}

assert_table <- function(table) {
  stopifnot(is.character(table), !anyNA(table), is.character(names(table)))
  
  dups <- names(table)[duplicated(names(table))]
  if (length(dups) > 0) {
    dups <- paste(dups, collapse = " ")
    stop(sprintf("Detected duplicated names (%s) in 'table'", dups))
  }
  
  if (all(nchar(table) == 0)) {
    return()
  }

  unknown <- setdiff(table, names(table))
  if (length(unknown) > 0) {
    missing <- paste(unknown, collapse = " ")
    stop(sprintf("Detected values (%s) in 'table' that are not in the names", missing))
  }
}

assert_anneal <- function(watson, crick, overhang, table = COMPLEMENT_TABLE_DNA) {
  assert_table(table)
  assert_in_alphabet(watson, alphabet = names(table))
  assert_in_alphabet(crick, alphabet = names(table))
  stopifnot(length(overhang) == 1, is.numeric(overhang), !is.na(overhang))
  stopifnot(-nchar(watson) < overhang, overhang < nchar(crick))

  crick_rc <- rc(crick, table = table)
  
  up <- substr(watson,   start = max(-overhang, 0) + 1, stop = nchar(crick)  - overhang)
  dn <- substr(crick_rc, start = max(+overhang, 0) + 1, stop = nchar(watson) + overhang)

  if (up != dn) {
    stop("Mismatched basepairs.")
  }
}


assert_checksum <- function(checksum, prefix = "") {
  stopifnot(length(checksum) == 1L, !is.na(checksum))
  if (nzchar(prefix)) {
    pattern <- "^(|(l|c)(s|d))seguid="
    stopifnot(grepl(pattern, checksum))
    checksum <- sub(pattern, "", checksum)
  }
  stopifnot(nchar(checksum) == 27)
}
